src/main/java/org/cactoos/experimental/Threads.java

```diff
@@ -23,13 +23,64 @@
  */
 package org.cactoos.experimental;
 
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import org.cactoos.Func;
+import org.cactoos.Scalar;
+import org.cactoos.collection.Mapped;
+import org.cactoos.iterable.IterableEnvelope;
+import org.cactoos.iterable.IterableOf;
+
 /**
  * Allows to execute the tasks concurrently.
  *
- * @param <X> The type of result objects.
- *
+ * @param <T> The type of task result item.
  * @since 1.0.0
+ * @todo #972:30min j.u.c.ExecutorService#invokeAll(java.util.Collection) should
+ *  be invoked with timeout. User should have the opportunity to pass the
+ *  timeout argument. We need a new decorator which supports the timeout.
+ * @todo #972:30min Add a new implementation which receive the tasks and able:
+ *  - create a new thread pool with predefined/specified size;
+ *  - execute tasks concurrently;
+ *  - shutdown the thread pool.
  */
-public interface Threads<X> extends Iterable<X> {
+public final class Threads<T> extends IterableEnvelope<T> {
+
+    /**
+     * Ctor.
+     * @param exc The executor.
+     * @param tasks The tasks to be executed concurrently.
+     */
+    @SafeVarargs
+    public Threads(final ExecutorService exc, final Scalar<T>... tasks) {
+        this(exc, new IterableOf<>(tasks));
+    }
 
+    /**
+     * Ctor.
+     * @param exc The executor.
+     * @param tasks The tasks to be executed concurrently.
+     * @checkstyle IllegalCatchCheck (20 lines)
+     */
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    public Threads(final ExecutorService exc, final Iterable<Scalar<T>> tasks) {
+        super(() -> {
+            try {
+                return new Mapped<>(
+                    Future::get,
+                    exc.invokeAll(
+                        new Mapped<>(
+                            (Func<Scalar<T>, Callable<T>>) task -> task::value,
+                            tasks
+                        )
+                    )
+                );
+            } catch (final Exception exp) {
+                Thread.currentThread().interrupt();
```

Comment:
@llorllale, however we will have another issue.
I can't simulate the InterruptedException, therefore this section can't be covered by unit test.
As the result, the code-coverage check will fail.
I tried in such way:
```java
    @Test(expected = UncheckedIOException.class)
    public void executionIsFailedDueToInterruptedThread() {
        new Threads<String>(
            Executors.newSingleThreadExecutor(),
            (Scalar<String>) () -> {
                TimeUnit.MILLISECONDS.sleep(50);
                Thread.currentThread().interrupt();
                throw new InterruptedException(
                    "The task was interrupted"
                );
            }
        ).iterator().next();
    }
```
and the coverage is 
![image](https://user-images.githubusercontent.com/1651114/50172902-1783ec00-031c-11e9-87f9-afdfee2e4e17.png)

As alternative, we can implement the `ExecutorService` which will throw the interrupted exception, but as you told, cactoos doesn't provide the `fakes` and such type of classes.

Could you please guide me regarding next steps here?

Reply:
@dgroup let's take a queue from [this](https://stackoverflow.com/a/36365763/1623885) stackoverflow answer and simply remove `Thread.currentThread().interrupt()`

Reply:
@llorllale, in case if we remove the `Thread.currentThread().interrupt()` then there is no sense to catch the InterruptedException at all as we will have the same handling procedure
```java
           } catch (final InterruptedException exp) {
                throw new CompletionException(exp);
            } catch (final Exception exp) {
                throw new CompletionException(exp);
            }
```

Reply:
@llorllale, done.